services/src/contexts/postgres.rs

```diff
@@ -0,0 +1,2597 @@
+use crate::api::model::datatypes::DatasetName;
+use crate::contexts::{ApplicationContext, QueryContextImpl, SessionId, SimpleSession};
+use crate::contexts::{GeoEngineDb, SessionContext};
+use crate::datasets::add_from_directory::{
+    add_datasets_from_directory, add_providers_from_directory,
+};
+use crate::datasets::upload::{Volume, Volumes};
+use crate::error::{self, Error, Result};
+use crate::layers::add_from_directory::{
+    add_layer_collections_from_directory, add_layers_from_directory, UNSORTED_COLLECTION_ID,
+};
+use crate::layers::storage::INTERNAL_LAYER_DB_ROOT_COLLECTION_ID;
+
+use crate::projects::{ProjectId, STRectangle};
+use crate::tasks::{SimpleTaskManager, SimpleTaskManagerBackend, SimpleTaskManagerContext};
+use async_trait::async_trait;
+use bb8_postgres::{
+    bb8::Pool,
+    bb8::PooledConnection,
+    tokio_postgres::{error::SqlState, tls::MakeTlsConnect, tls::TlsConnect, Config, Socket},
+    PostgresConnectionManager,
+};
+use geoengine_datatypes::raster::TilingSpecification;
+use geoengine_operators::engine::ChunkByteSize;
+use geoengine_operators::util::create_rayon_thread_pool;
+use log::{debug, info};
+use postgres_protocol::escape::escape_literal;
+use rayon::ThreadPool;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use super::{ExecutionContextImpl, Session, SimpleApplicationContext};
+
+// TODO: do not report postgres error details to user
+
+/// A contex with references to Postgres backends of the dbs. Automatically migrates schema on instantiation
+#[derive(Clone)]
+pub struct PostgresContext<Tls>
+where
+    Tls: MakeTlsConnect<Socket> + Clone + Send + Sync + 'static,
+    <Tls as MakeTlsConnect<Socket>>::Stream: Send + Sync,
+    <Tls as MakeTlsConnect<Socket>>::TlsConnect: Send,
+    <<Tls as MakeTlsConnect<Socket>>::TlsConnect as TlsConnect<Socket>>::Future: Send,
+{
+    default_session_id: SessionId,
+    thread_pool: Arc<ThreadPool>,
+    exe_ctx_tiling_spec: TilingSpecification,
+    query_ctx_chunk_size: ChunkByteSize,
+    pub(crate) task_manager: Arc<SimpleTaskManagerBackend>,
+    pub(crate) pool: Pool<PostgresConnectionManager<Tls>>,
```

Comment:
Why don't we create two pub(crate) accessor functions?
We don't make the other fields public.

Reply:
7307aba6e093811fce9d92d075c4314afd7b1e03